### PR TITLE
CGM dN/dX

### PR DESCRIPTION
**Pull request by profxj**
Simple dN/dX calculation for the CGM

### CODE REVIEW

**Comment by profxj** [comment]
Merging (usualy Python 3.5 fail)

**Comment by profxj** [comment]
Check that.  Waiting for Burchett to approve

File: pyigm/cgm/analysis.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    """ Estimate dN/dX for a set of CGM assuming unit covering fraction
+    Use beta=0 and rvir_Lstar=300kpc for a constant CGM to 300kpc
+
+    Parameters
+    ----------
+    Lrng : tuple, optional
+      Range to integrate luminosity in terms of L*
+    beta : float, optional
+      Parameterization of rvir with L
+       r_vir = 250 kpc * (L/L*)^beta
```
Comment (jnburchett) [inline pyigm/cgm/analysis.py:30]: Per Prochaska+11

File: pyigm/cgm/analysis.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    phi_str, alpha, Mstar : Quantity, float, float
```
Comment (jnburchett) [inline pyigm/cgm/analysis.py]: Should indicate somewhere that phi_str has a factor of h<sup>3</sup> in it as well.

File: pyigm/cgm/analysis.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import pdb
+
+from scipy.special import gamma, gammainc
+
+from astropy import units as u
+from astropy import constants as const
+from astropy.cosmology import Planck15
+
+def dndx_rvir(Lrng=(0.001, 10), beta=0.2, rvir_Lstar=250.*u.kpc,
+         phi_str = 1.49/u.Mpc**3, alpha = -1.05, Mstar = -20.44,
```
Comment (jnburchett) [inline pyigm/cgm/analysis.py]: Blanton gives the value for phi_str as 1.49 x 10<sup>-2</sup> h<sup>3</sup>.  Line 50 includes a factor of 10<sup>-2</sup>; coincidence?

File: pyigm/cgm/analysis.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    dNdX : float
+      Cumulative dNdX
+
+    """
+    # Cosmology
+    if cosmo is None:
+        cosmo = Planck15
+    hubb = cosmo.H0.value / 100.
+    # Constants
+    phi_str_cgs = (phi_str * 1e-2 * hubb)
```
Comment (jnburchett) [inline pyigm/cgm/analysis.py]: Not following this conversion.  I get a huge factor of 10<sup>-74</sup> to get to cgs?!

File: pyigm/cgm/analysis.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    dndx_const = (const.c / cosmo.H0).cgs
```
Comment (jnburchett) [inline pyigm/cgm/analysis.py:54]: Is the actual unit conversion dirty work somehow handled here?

File: pyigm/cgm/analysis.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    # CGM extends to rvir with power-law dependence
```
Comment (jnburchett) [inline pyigm/cgm/analysis.py]: Seems like a weird place for this comment.  Do you mean something like 'CGM extends to rvir; parameterize rvir as power law of L/L*' ?

File: pyigm/cgm/analysis.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    Lval = np.linspace(Lrng[0], Lrng[1], 1000)
+    x = alpha + 1 + beta
+    # Integrate
+    dNdx_rvir = (dndx_const * phi_str_cgs * (np.pi * rvir_Lstar**2) * (
+        gamma(x) * ( gammainc(x,Lrng[1]) - gammainc(x,Lval)))).decompose()
```
Comment (jnburchett) [inline pyigm/cgm/analysis.py]: That decompose function is nifty...hadn't seen that one.  However, I'm worried about the phi_str_cgs factor still, as I don't totally get the derivation of this expression.

File: pyigm/cgm/analysis.py (hunk 1 above)
Comment (profxj) [inline pyigm/cgm/analysis.py:30]: added

File: pyigm/cgm/analysis.py (hunk 2 above)
Comment (profxj) [inline pyigm/cgm/analysis.py]: Good catch.  I forgot the h**3 !

File: pyigm/cgm/analysis.py (hunk 3 above)
Comment (profxj) [inline pyigm/cgm/analysis.py]: As above.  Corrected now

File: pyigm/cgm/analysis.py (hunk 4 above)
Comment (profxj) [inline pyigm/cgm/analysis.py]: This use of variable name was historic (IDL port).

The Mpc**3 is in the def line.  Improving this..

File: pyigm/cgm/analysis.py (hunk 5 above)
Comment (profxj) [inline pyigm/cgm/analysis.py:54]: Not really.  That is done with the decompose below

File: pyigm/cgm/analysis.py (hunk 6 above)
Comment (profxj) [inline pyigm/cgm/analysis.py]: Removed.

**Comment by profxj** [comment]
Thanks for all these comments JB.

**Pull request merged**